src/operator/contrib/bounding_box-inl.h

```diff
@@ -0,0 +1,730 @@
+    Tensor<xpu, 3, DType> out = outputs[box_nms_enum::kOut]
+     .get_with_shape<xpu, 3, DType>(Shape3(num_batch, num_elem, width_elem), s);
+    Tensor<xpu, 3, DType> record = outputs[box_nms_enum::kTemp]
+     .get_with_shape<xpu, 3, DType>(Shape3(num_batch, num_elem, 1), s);
+
+    // prepare workspace
+    Shape<1> sort_index_shape = Shape1(num_batch * num_elem);
+    Shape<3> buffer_shape = Shape3(num_batch, num_elem, width_elem);
+    index_t workspace_size = 4 * sort_index_shape.Size();
+    if (req[0] == kWriteInplace) {
+      workspace_size += buffer_shape.Size();
+    }
+    Tensor<xpu, 1, DType> workspace = ctx.requested[box_nms_enum::kTempSpace]
+      .get_space_typed<xpu, 1, DType>(Shape1(workspace_size), s);
+    Tensor<xpu, 1, DType> sorted_index(workspace.dptr_, sort_index_shape, s);
+    Tensor<xpu, 1, DType> scores(sorted_index.dptr_ + sorted_index.MSize(),
+      sort_index_shape, s);
+    Tensor<xpu, 1, DType> batch_id(scores.dptr_ + scores.MSize(), sort_index_shape,
+      s);
+    Tensor<xpu, 1, DType> areas(batch_id.dptr_ + batch_id.MSize(), sort_index_shape, s);
+    Tensor<xpu, 3, DType> buffer = data;
+    if (req[0] == kWriteInplace) {
+      // make copy
+      buffer = Tensor<xpu, 3, DType>(areas.dptr_ + areas.MSize(), buffer_shape, s);
+      buffer = F<mshadow_op::identity>(data);
+    }
+
+    // indecies
+    int score_index = param.score_index;
+    int coord_start = param.coord_start;
+    int id_index = param.id_index;
+
+    // sort topk
+    int topk = param.topk < 0? num_elem : std::min(num_elem, param.topk);
+    if (topk < 1) {
+      out = F<mshadow_op::identity>(buffer);
+      record = reshape(range<DType>(0, num_batch * num_elem), record.shape_);
+      return;
+    }
+    scores = reshape(slice<2>(buffer, score_index, score_index + 1), scores.shape_);
+    sorted_index = range<DType>(0, num_batch * num_elem);
+    mxnet::op::SortByKey(scores, sorted_index, false);
+    batch_id = F<mshadow_op::floor>(sorted_index / ScalarExp<DType>(num_elem));
+    mxnet::op::SortByKey(batch_id, scores, true);
+    batch_id = F<mshadow_op::floor>(sorted_index / ScalarExp<DType>(num_elem));
+    mxnet::op::SortByKey(batch_id, sorted_index, true);
+
+    // pre-compute areas of candidates
+    areas = 0;
+    Kernel<compute_area, xpu>::Launch(s, num_batch * topk, areas.dptr_,
+     buffer.dptr_ + coord_start, sorted_index.dptr_, topk, num_elem, width_elem,
+     param.in_format);
+
+    // apply nms
+    // go through each box as reference, suppress if overlap > threshold
+    // sorted_index with -1 is marked as suppressed
+    for (int ref = 0; ref < topk; ++ref) {
+      int num_worker = topk - ref - 1;
+      if (num_worker < 1) continue;
+      Kernel<nms_impl, xpu>::Launch(s, num_batch * num_worker, sorted_index.dptr_,
+        buffer.dptr_, areas.dptr_, num_worker, ref, num_elem, width_elem,
+        coord_start, id_index, param.overlap_thresh, param.force_suppress, param.in_format);
+    }
```

Comment:
Does it mean that we need to launch the kernel for `topk` times?

Reply:
Yep, that is because we need to consider broadcasting suppression problem.
If scores A > B > C, A suppress B with IOU 0.8, B suppress C with IOU 0.6, A don't suppress C with IOU 0.4. Then we should suppress B and keep C because B is already suppress by A and C has nothing to do with A.
Therefore, until we get the status of B, we can't determine the status of C.

Reply:
OK.  There may be a possible speed problem if `topk` is large. However, the current implementation should be good in order to run the experiments. Could we later revise the kernel using similar logic as https://github.com/rbgirshick/py-faster-rcnn/blob/master/lib/nms/nms_kernel.cu?

Reply:
Marked. I can later tackle on this part, for now I don't encounter perf issue with normal settings.